File: src/main/java/com/telegram/codex/integration/telegram/application/webhook/InboundMessageProcessor.java
package com.telegram.codex.integration.telegram.application.webhook;

import com.telegram.codex.conversation.application.JobSchedulerService;
import com.telegram.codex.conversation.infrastructure.MediaGroupBufferRepository;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import com.telegram.codex.integration.telegram.domain.webhook.TelegramUpdate;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.List;

@Component
public class InboundMessageProcessor {

    private final UnsupportedMessageHandler unsupportedMessageHandler;
    private final DuplicateUpdateHandler duplicateUpdateHandler;
    private final TelegramCommandHandler telegramCommandHandler;
    private final ReplyRequestGuard replyRequestGuard;
    private final MediaGroupBufferRepository mediaGroupStore;
    private final JobSchedulerService jobSchedulerService;
    private final List<ProcessingStep> processingSteps;

    public InboundMessageProcessor(
        UnsupportedMessageHandler unsupportedMessageHandler,
        DuplicateUpdateHandler duplicateUpdateHandler,
        TelegramCommandHandler telegramCommandHandler,
        ReplyRequestGuard replyRequestGuard,
        MediaGroupBufferRepository mediaGroupStore,
        JobSchedulerService jobSchedulerService
    ) {
        this.unsupportedMessageHandler = unsupportedMessageHandler;
        this.duplicateUpdateHandler = duplicateUpdateHandler;
        this.telegramCommandHandler = telegramCommandHandler;
        this.replyRequestGuard = replyRequestGuard;
        this.mediaGroupStore = mediaGroupStore;
        this.jobSchedulerService = jobSchedulerService;
        // 順序好重要：先擋 unsupported / duplicate，再處理 command，最後先落一般 reply flow。
        this.processingSteps = List.of(
            this::handleUnsupportedMessage,
            this::handleDuplicateUpdate,
            this::handleTelegramCommand,
            this::handleBlockedReplyRequest
        );
    }

    public void process(InboundMessage message) {
        process(message, null);
    }

    public void process(InboundMessage message, TelegramUpdate update) {
        if (handledByProcessingStep(message, update)) {
            return;
        }
        jobSchedulerService.enqueueReplyGeneration(message);
    }

    public void deferMediaGroup(InboundMessage message, Duration waitDuration) {
        MediaGroupBufferRepository.EnqueueResult result = mediaGroupStore.enqueue(message, waitDuration.toMillis() / 1000.0);
        jobSchedulerService.scheduleMediaGroupFlush(result.key(), result.deadlineAt(), waitDuration);
    }

    private boolean handledByProcessingStep(InboundMessage message, TelegramUpdate update) {
        for (ProcessingStep step : processingSteps) {
            if (step.handle(message, update)) {
                return true;
            }
        }
        return false;
    }

    private boolean handleUnsupportedMessage(InboundMessage message, TelegramUpdate update) {
        return unsupportedMessageHandler.handle(message, update);
    }

    private boolean handleDuplicateUpdate(InboundMessage message, TelegramUpdate update) {
        return duplicateUpdateHandler.handle(message);
    }

    private boolean handleTelegramCommand(InboundMessage message, TelegramUpdate update) {
        return telegramCommandHandler.handle(message);
    }

    private boolean handleBlockedReplyRequest(InboundMessage message, TelegramUpdate update) {
        return !replyRequestGuard.allow(message);
    }

    @FunctionalInterface
    private interface ProcessingStep {
        boolean handle(InboundMessage message, TelegramUpdate update);
    }
}

File: src/telegram/inbound/inbound_message_processor.rs
use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;

use crate::config::AppConfig;
use crate::conversation::scheduler::job_scheduler::{InboundMessageProcessorPort, JobScheduler};
use crate::conversation::storage::media_group_buffer_repository::MediaGroupBufferRepository;
use crate::telegram::commands::telegram_command_handler::TelegramCommandHandler;
use crate::telegram::inbound::duplicate_update_handler::DuplicateUpdateHandler;
use crate::telegram::inbound::reply_request_guard::ReplyRequestGuard;
use crate::telegram::inbound::unsupported_message_handler::UnsupportedMessageHandler;
use crate::telegram::shared::inbound_message::InboundMessage;
use crate::telegram::shared::telegram_types::TelegramUpdate;

pub struct InboundMessageProcessor {
    unsupported_message_handler: Arc<UnsupportedMessageHandler>,
    duplicate_update_handler: Arc<DuplicateUpdateHandler>,
    telegram_command_handler: Arc<TelegramCommandHandler>,
    reply_request_guard: Arc<ReplyRequestGuard>,
    media_group_store: Arc<MediaGroupBufferRepository>,
    job_scheduler: Arc<JobScheduler>,
    config: Arc<AppConfig>,
}

impl InboundMessageProcessor {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        unsupported_message_handler: Arc<UnsupportedMessageHandler>,
        duplicate_update_handler: Arc<DuplicateUpdateHandler>,
        telegram_command_handler: Arc<TelegramCommandHandler>,
        reply_request_guard: Arc<ReplyRequestGuard>,
        media_group_store: Arc<MediaGroupBufferRepository>,
        job_scheduler: Arc<JobScheduler>,
        config: Arc<AppConfig>,
    ) -> Self {
        Self {
            unsupported_message_handler,
            duplicate_update_handler,
            telegram_command_handler,
            reply_request_guard,
            media_group_store,
            job_scheduler,
            config,
        }
    }

    /// Runs the inbound pipeline: unsupported -> duplicate -> command -> guard,
    /// and only then queues an actual reply.
    pub async fn process(&self, message: Option<&InboundMessage>, update: Option<&TelegramUpdate>) -> Result<()> {
        let Some(message) = message else {
            self.unsupported_message_handler.handle(None, update).await?;
            return Ok(());
        };
        if self.unsupported_message_handler.handle(Some(message), update).await? {
            return Ok(());
        }
        if self.duplicate_update_handler.handle(message).await? {
            return Ok(());
        }
        if self.telegram_command_handler.handle(message).await? {
            return Ok(());
        }
        if !self.reply_request_guard.allow(message).await? {
            return Ok(());
        }
        self.job_scheduler.enqueue_reply_generation(message.clone());
        Ok(())
    }

    /// Albums arrive as separate updates, so they are buffered until the group
    /// is complete.
    pub async fn defer_media_group(&self, message: &InboundMessage) -> Result<()> {
        let (result, key) = self.media_group_store.enqueue(message, self.config.media_group_wait_ms as f64 / 1000.0).await?;
        self.job_scheduler.schedule_media_group_flush(key, result.deadline_at, self.config.media_group_wait_ms);
        Ok(())
    }
}

#[async_trait]
impl InboundMessageProcessorPort for InboundMessageProcessor {
    async fn process(&self, message: Option<InboundMessage>) -> Result<()> {
        InboundMessageProcessor::process(self, message.as_ref(), None).await
    }
}
